node/engine/chainservice/eth_chainservice.go

```diff
@@ -335,22 +344,42 @@ out:
 			ecs.logger.Trace().Msg("resubscribed to new blocks")
 
 		case newBlock := <-newBlockChan:
-			ecs.logger.Debug().Msgf("detected new block: %d", newBlock.Number.Uint64())
-			// When we get a new block, check the events at the top of the queue.
-			// If they have enough block confirmations, remove them from the queue and process them
-			for eventQueue.Len() > 0 && newBlock.Number.Uint64() >= (*eventQueue)[0].BlockNumber+REQUIRED_BLOCK_CONFIRMATIONS {
-				chainEvent := heap.Pop(eventQueue).(ethTypes.Log)
-				ecs.logger.Debug().Msgf("event popped from queue (updated queue length: %d)", eventQueue.Len())
-				err := ecs.dispatchChainEvents([]ethTypes.Log{chainEvent})
-				if err != nil {
-					errorChan <- fmt.Errorf("failed dispatchChainEvents: %w", err)
-					break out
-				}
-			}
+			newBlockNum := newBlock.Number.Uint64()
+			ecs.logger.Debug().Msgf("detected new block: %d", newBlockNum)
+			ecs.updateEventTracker(errorChan, &newBlockNum, nil)
+
+			ecs.logger.Debug().Msgf("detected new block: %d", newBlockNum)
 		}
 	}
 }
 
+// updateEventTracker accepts a new block number and/or new event and dispatches a chain event if there are enough block confirmations
+func (ecs *EthChainService) updateEventTracker(errorChan chan<- error, blockNumber *uint64, chainEvent *ethTypes.Log) {
+	// lock the mutex for the shortest amount of time. The mutex only need to be locked to update the eventTracker data structure
+	ecs.eventTracker.mu.Lock()
```

Comment:
Instead of protecting calls to the `EventQueue` with a mutex in the chain service, could the `EventQueue` use mutexs (or some other `sync` feature) internally to protect against concurrent calls to `Pop`/`Push`? Kind of like [our safesync.Map](https://github.com/statechannels/go-nitro/blob/638e0447c56448cdcfe6fc1ff2a8b2f255963964/internal/safesync/safesync.go#L7)

Reply:
Definitely an option! One nice feature of the current implementation is that the mutex also protects latest block variable read and writes.